### PR TITLE
Added client metadata flag, fixes #3091

### LINKED ISSUE #3091: CLI `clients create` command is missing flag to set Metadata
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/hydra/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/hydra/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Cloud](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

When creating clients from the CLI, it's not currently possible to set the `OAuth2Client.Metadata` property. 

### Reproducing the bug

1. Run `hydra clients create --help` and check the flags.

### Relevant log output

_No response_

### Relevant configuration

_No response_

### Version

Latest

### On which operating system are you observing this issue?

_No response_

### In which environment are you deploying?

Binary

### Additional Context

I earlier submitted https://github.com/ory/x/pull/497 which will allow setting `OAuth2Client.Metadata` to a `map[string]string` value, which works well for my use case. However looking at the API reference it seems that metadata is a `JSONRawMessage` or `interface{}`, so should the CLI permit setting `OAuth2Client.Metadata` to be arbitrary JSON as a string input, or does a `map[string]string` suffice for most CLI use cases?

If `map[string]string` is sufficient I'll submit a PR to add that. Thanks.

### PR DESCRIPTION
**Pull request by rikkuness**
Adds a `--metadata` flag to both `hydra clients create` and `hydra clients update`. Also ensures that valid JSON is being passed.

## Related issue(s)

Fixes #3091

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you! Can you please add this field to the tests at: https://github.com/ory/hydra/blob/master/cmd/root_test.go? :)

**Comment by rikkuness** [comment]
Sorry about that, I missed that file the first time around 😄 

Tests added, I tried to add a test to test for invalid JSON with the `expectErr` in the tests but as the CLI is using `cmdx.Must()` which calls an `os.Exit()`, the `rootCmd.Execute()` never actually returns an error so you can't test for it.

File: cmd/cli/handler_client.go [hunk 1]
```diff
@@ -101,6 +101,13 @@ func (h *ClientHandler) CreateClient(cmd *cobra.Command, args []string) {
 	ek, encryptSecret, err := newEncryptionKey(cmd, nil)
 	cmdx.Must(err, "Failed to load encryption key: %s", err)
 
+	var metadata models.JSONRawMessage
+	metadataStr := flagx.MustGetString(cmd, "metadata")
+	if metadataStr != "" {
+		err = json.Unmarshal([]byte(flagx.MustGetString(cmd, "metadata")), &metadata)
+		cmdx.Must(err, "Failed to parse metadata: %s", err)
```
Comment (aeneasr) [inline cmd/cli/handler_client.go:108]: Can you please use https://github.com/tidwall/gjson/blob/master/gjson.go#L2549 instead and improve the error message a bit? :)

**Review by aeneasr — CHANGES_REQUESTED** [review]

File: cmd/cli/handler_client.go (hunk 1 above)
Comment (rikkuness) [reply cmd/cli/handler_client.go:108]: Sure :) Although the JSON string, will it still need to also be parsed out to type `JSONRawMessage` to pass into the `OAuth2Client` struct?

Comment (aeneasr) [reply cmd/cli/handler_client.go:108]: Yes, but `JSONRawMessage` is just `[]byte()`, so:

```
var JSONRawMessage = []byte(`{"foo":"bar"}`)
```

Comment (rikkuness) [reply cmd/cli/handler_client.go:108]: So if I set `Metadata` as above as just a raw string from the CLI input, then it gets added to the client as: 
```json
"metadata": "{\"foo\":\"bar\"}",
```
Which is not what I would expect to see. However if I do something like:
```go
var metadata interface{}
json.Unmarshal([]byte(flagx.MustGetString(cmd, "metadata")), &metadata)
```

Then in the resulting client output I get:
```json
"metadata": {
	"foo": "bar"
},
```

Which is much more like what I was expecting

Comment (aeneasr) [reply cmd/cli/handler_client.go:108]:
```
var metadata json.RawMessage = []byte(flagx.MustGetString(cmd, "metadata"))
```

is equivalent and to much faster than

```
var metadata interface{}
json.Unmarshal([]byte(flagx.MustGetString(cmd, "metadata")), &metadata)
```

:)

Comment (rikkuness) [reply cmd/cli/handler_client.go:108]: Ohhhh, my bad, yeah I was missing the fact `JSONRawMessage` implemented an Unmarshaler! 

I'll update that now, thanks!

Comment (aeneasr) [reply cmd/cli/handler_client.go:108]: awesome :)

Comment (rikkuness) [reply cmd/cli/handler_client.go]: Apologies, I'd made some local changes and then completely forgot about this PR 😬 I've just pushed now, does that look more like what you'd expect? It seemed long winded but `gjson.Valid` returned a bool so I wasn't sure the preferred way to error out, at the moment it's just printing there was an error and carrying on anyway

**Comment by rikkuness** [comment]
Hey, is this superseded by the recent commits improve `create client`? Looks like those commits closed the original issue

**Comment by aeneasr** [comment]
Yes!

**Comment by rikkuness** [comment]
Awesome, thanks!

**Pull request closed without merging**